File: CS50_Computer_Science/sentimental-readability/readability.py
import re
from cs50 import get_string

def count_letters(text):
    letters = sum(1 for char in text if char.isalpha())
    return letters

def count_words(text):
    words = len(text.split())
    return words

def count_sentences(text):
    sentences = sum(1 for char in text if char in '.!?')
    return sentences

def calculate_index(letters, words, sentences):
    L = (letters / words) * 100
    S = (sentences / words) * 100
    index = 0.0588 * L - 0.296 * S - 15.8
    return round(index)

def main():
    text = get_string("Text: ")
    letters = count_letters(text)
    words = count_words(text)
    sentences = count_sentences(text)
    index = calculate_index(letters, words, sentences)

    if index < 1:
        print("Before Grade 1")
    elif index >= 16:
        print("Grade 16+")
    else:
        print(f"Grade {index}")

if __name__ == "__main__":
    main()

File: CS50_Computer_Science/readability/readability.c
#include <ctype.h>
#include <cs50.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

// Prototipos de funciones
int count_letters(string text);
int count_words(string text);
int count_sentences(string text);

int main(void)
{
    // Solicitar al usuario que ingrese un texto
    string text = get_string("Text: ");

    // Contar el número de letras, palabras y oraciones en el texto
    int letters = count_letters(text);
    int words = count_words(text);
    int sentences = count_sentences(text);

    // Calcular el índice de Coleman-Liau
    float L = ((float) letters / words) * 100;
    float S = ((float) sentences / words) * 100;
    float index = 0.0588 * L - 0.296 * S - 15.8;

    // Redondear el índice al número entero más cercano
    int grade = round(index);

    // Imprimir el nivel de grado correspondiente
    if (grade < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (grade >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", grade);
    }
}

// Función para contar el número de letras en el texto
int count_letters(string text)
{
    int letters = 0;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (isalpha(text[i]))
        {
            letters++;
        }
    }
    return letters;
}

// Función para contar el número de palabras en el texto
int count_words(string text)
{
    int words = 1; // Comenzamos en 1 porque la última palabra no tiene un espacio después
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (isspace(text[i]))
        {
            words++;
        }
    }
    return words;
}

// Función para contar el número de oraciones en el texto
int count_sentences(string text)
{
    int sentences = 0;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            sentences++;
        }
    }
    return sentences;
}
